planner/core/logical_plan_builder.go

```diff
@@ -576,6 +576,23 @@ func (b *PlanBuilder) buildProjectionField(id, position int, field *ast.SelectFi
 	}
 }
 
+func eliminateIfNullOnNonNullColumn(p LogicalPlan, expr expression.Expression) expression.Expression {
+	if scalarExpr, ok := expr.(*expression.ScalarFunction); ok {
```

Comment:
The code nesting is too deep, how about:
```go
ds, isDs := p.(*DataSource)
if !isDs {
	return expr
}

scalarExpr, isScalarFunc := expr.(*expression.ScalarFunction);
if !isScalarFunc || !scalarExpr.FuncName.L == ast.Ifnull {
	return expr
}

colRef, isColRef := scalarExpr.GetArgs()[0].(*expression.Column)
if !isColRef {
	return expr
}

colInfo := model.FindColumnInfo(ds.Columns, cExpr.ColName.L)
if !mysql.HasNotNullFlag(colFound.Flag) {
	return expr
}

return colRef
```

Reply:
another thing, should we recursively check the expression tree to eliminate as much `IfNull` as possible?

Reply:
for example:

```sql
select a + ifnull(b, 1) from t;
```